### PR TITLE
feat: Use reference equality for `Symbol`

### PR DESCRIPTION
**Pull request by Marwes**
Rather than using equality by name we can compare `Symbol` by the reference that it points which lets us immediately distinguish between the symbols of two bindings.

```
(x) => {
    // These two `x` below are not the same as the function parameter due shadowing
    x = 1
    return x
}
```

This gives us two benefits, first any lookups done with `Symbol` keys are faster as it only needs to compare/hash a single pointer instead of the full name. Second, in any analysis done after `convert` we don't need to care about scope to the same extent as two symbols comparing equal guarantees that they refer to the same thing which would simplify getting all references to a variable in the LSP (for instance).

This alters the flatbuffer serialization and deserialization to use (and rely on) shared strings so that this reference equality is preserved across flatbuffers at least on the rust side, the go side does not appear to need this at the moment.

Based on #4376

### CODE REVIEW

**Comment by Marwes** [comment]
That it compares references here are just a convenient way to have cheap equality checks and also store the actual string, if we didn't need the string stored directly in `Symbol` we could just have `Symbol(u32)` if we wanted.

**Review by skartikey — APPROVED** [review]
The change looks nice to me, O(1) equality check using `Symbol` would definitely be good for performance gain. 
I have a question on the cloning of the reference below.
I am approving it but as it's a huge change I would like someone in the team to have a look as well.

File: libflux/flux-core/src/semantic/convert.rs [hunk 1]
```diff
@@ -1226,99 +1340,53 @@ mod tests {
...
-            package: "main".to_string(),
+            package: "foo".to_string(),
             files: vec![File {
                 loc: b.location.clone(),
                 package: Some(PackageClause {
                     loc: b.location.clone(),
                     name: Identifier {
                         loc: b.location.clone(),
-                        name: Symbol::from("foo"),
+                        name: symbols["foo"].clone(),
```
[the remark this answers is not in the document]
Comment (Marwes) [reply libflux/flux-core/src/semantic/convert.rs:1357]: `Symbol` contains an `Arc` which is the pointer that we compare. So `symbol == symbol.clone()` is always true`.

**Review by scbrickley — APPROVED** [review]
I don't see anything that needs changing in this patch, but I am not very familiar with this part of the code base. Someone else on the team may be able to give a more in-depth review if you think you need it.

**Pull request merged**